utils/near-performance-metrics/src/stats.rs

```diff
@@ -0,0 +1,341 @@
+            if *x.borrow_mut() == 0 {
+                *x.borrow_mut() = NTHREADS.fetch_add(1, Ordering::SeqCst);
+                self.stats.insert(*x.borrow_mut(), ThreadStats::new());
+            }
+            *x.borrow_mut()
+        });
+        if let Some(val) = self.stats.get_mut(&tid) {
+            val.log(class_name, msg, line, took)
+        }
+    }
+
+    fn print_stats(&self, sleep_time: Duration) {
+        info!("Performance stats {} threads", self.stats.len());
+        for entry in self.stats.iter() {
+            entry.1.print_stats(*entry.0, sleep_time);
+        }
+    }
+}
+
+pub fn measure_performance<F, Message, Result>(
+    class_name: &'static str,
+    msg: Message,
+    f: F,
+) -> Result
+where
+    F: FnOnce(Message) -> Result,
+{
+    let now = Instant::now();
+    let result = f(msg);
+
+    let took = now.elapsed();
+    if took > SLOW_CALL_THRESHOLD {
+        info!(
+            "SLOW {}:{} {:?} took: {}ms",
+            class_name,
+            TID.with(|x| *x.borrow()),
+            std::any::type_name::<Message>(),
+            took.as_millis(),
+        );
+    }
+
+    STATS.write().unwrap().log(class_name, std::any::type_name::<Message>(), 0, took);
+    result
+}
+
+pub fn measure_performance2<F, Result>(
+    class_name: &'static str,
+    msg_name: &'static str,
+    f: F,
+) -> Result
+where
+    F: FnOnce() -> Result,
+{
+    let now = Instant::now();
+    let result = f();
+
+    let took = now.elapsed();
+    if took > SLOW_CALL_THRESHOLD {
+        info!(
+            "SLOW {}:{} {:?} took: {}ms",
+            class_name,
+            TID.with(|x| *x.borrow()),
+            msg_name,
+            took.as_millis(),
+        );
+    }
+
+    STATS.write().unwrap().log(class_name, msg_name, 0, took);
+    result
+}
+
+#[allow(unused_variables)]
+pub fn performance_with_debug<F, Message, Result>(
+    class_name: &'static str,
+    msg: Message,
+    f: F,
+) -> Result
+where
+    F: FnOnce(Message) -> Result,
+    Message: Debug,
+{
+    #[cfg(not(feature = "performance_stats"))]
+    {
+        f(msg)
+    }
+
+    #[cfg(feature = "performance_stats")]
+    {
+        let now = Instant::now();
+        let msg_test = format!("{:?}", msg);
+        let result = f(msg);
+
+        let took = now.elapsed();
+        if took > SLOW_CALL_THRESHOLD {
+            let msg_length = msg_test.len();
+            let mut msg_test = msg_test;
+            if msg_length >= MESSAGE_LIMIT {
+                msg_test = msg_test.as_str()[..MESSAGE_LIMIT].to_string();
+            }
+            info!(
+                "SLOW {}:{} {:?} took: {}ms len: {} {}",
+                class_name,
+                TID.with(|x| *x.borrow()),
+                std::any::type_name::<Message>(),
+                took.as_millis(),
+                msg_length,
+                msg_test
+            );
+        }
+
+        STATS.write().unwrap().log(class_name, std::any::type_name::<Message>(), 0, took);
```

Comment:
Is lock contention a concern here? What a large number of actors try to write at the same time

Reply:
That should not be an issue, we don't many messages.